src/Bicep.Core/LanguageConstants.cs

```diff
@@ -73,6 +73,39 @@ public static class LanguageConstants
         public const string ResourceDependsOnPropertyName = "dependsOn";
         public const string TypeNameString = "string";
 
+
+        /*
+         * The following top-level properties are deploy-time constant string,
+         * but unlike name, apiVersion and type, they might be absent from resource bodies.
+         */
+        public static readonly string[] OptionalDeployTimeConstantPropertyNames = new[]
+        {
+            "location",
+            "kind",
+            "subscriptionId",
+            "resourceGroup",
+            "managedBy",
```

Comment:
TBH, although it generally should be safe to do this, I'd be happy keeping this list even more restrictive (or possibly empty!). IMO the lack of validation (which this PR solves) is the bigger problem - assigning a temporary variable isn't too much trouble as long as you are aware that you need to do it (via validation).

I feel like even having `location` in this list is worth a DSL discussion topic - RPs **shouldn't** change this value, but I know that some do - to me it feels safer to just avoid the problem altogether.

Reply:
I'm sold! I think I'm more inclined to make the list empty now. Let's hold off merging the PR and have a DSL discussion next Monday.

PS: there are a lot of RPs that do casing normalization for location, kind, resource IDs and enums, although it violates ARM RPC, and it hurts consistency. Most of the What-If KPI exception requests that I received are for casing normalization 😥.

Reply:
Removed the list to block all optional top-level properties as we discussed in today's DSL meeting.